src/h265_sei_parser.cc

```diff
@@ -0,0 +1,123 @@
+/*
+ *  Copyright (c) Facebook, Inc. and its affiliates.
+ */
+
+#include "h265_sei_parser.h"
+
+#include <stdio.h>
+
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include "h265_common.h"
+
+namespace h265nal {
+
+// General note: this is based off the 2016/12 version of the H.265 standard.
+// You can find it on this page:
+// http://www.itu.int/rec/T-REC-H.265
+
+// Unpack RBSP and parse SEI state from the supplied buffer.
+std::unique_ptr<H265SeiParser::SeiState> H265SeiParser::ParseSei(
+    const uint8_t* data, size_t length) noexcept {
+  std::vector<uint8_t> unpacked_buffer = UnescapeRbsp(data, length);
+  rtc::BitBuffer bit_buffer(unpacked_buffer.data(), unpacked_buffer.size());
+  return ParseSei(&bit_buffer);
+}
+
+std::unique_ptr<H265SeiParser::SeiState> H265SeiParser::ParseSei(
+    rtc::BitBuffer* bit_buffer) noexcept {
+  // H265 SEI NAL Unit (access_unit_delimiter_rbsp()) parser.
+  // Section 7.3.5 ("Supplemental enhancement information message syntax") of the H.265
+  // standard for a complete description.
+
+  uint32_t payload_type = 0;
+  uint32_t ff_byte = 0xff;
+  
+  while(ff_byte == 0xff) {
+    bit_buffer->ReadBits(8, ff_byte); // f(8)
+    payload_type += ff_byte;
+  }
+
+  std::unique_ptr<SeiState> sei = nullptr;
+  if(static_cast<SeiType>(payload_type) == SeiType::user_data_registered_itu_t_t35) {
+    sei = std::make_unique<SeiStateUserData>();
+  } else {
+    sei = std::make_unique<SeiStateNotImplemented>();
+  }
+  sei->payload_type = static_cast<SeiType>(payload_type);
+
+  ff_byte = 0xff;
+  while(ff_byte == 0xff) {
+    bit_buffer->ReadBits(8, ff_byte); // f(8)
+    sei->payload_size += ff_byte;
+  }
```

Comment:
I'd move this block to just line 43, just after the payload_type calculation. Both payload_size and payload_type should be treated the same. I'd suggest making them variables of the SeiState class (right now you only have payload_type)

Reply:
Done